src/NLU.DevOps.Luis.Shared/LuisNLUBatchTestClientBase.cs

```diff
@@ -0,0 +1,296 @@
+        public bool IsBatchEnabled { get; }
+
+        private ILuisBatchTestClient LuisBatchTestClient { get; }
+
+        /// <inheritdoc />
+        public abstract Task<ILabeledUtterance> TestAsync(JToken query, CancellationToken cancellationToken);
+
+        /// <inheritdoc />
+        public abstract Task<ILabeledUtterance> TestSpeechAsync(string speechFile, JToken query, CancellationToken cancellationToken);
+
+        /// <inheritdoc />
+        public async Task<IEnumerable<ILabeledUtterance>> TestAsync(IEnumerable<JToken> queries, CancellationToken cancellationToken)
+        {
+            if (queries == null)
+            {
+                throw new ArgumentNullException(nameof(queries));
+            }
+
+            // Chunk query set into batch sizes suitable for the LUIS batch evaluation endpoint
+            var queryBatches = queries
+                .Select(ToLuisBatchInput)
+                .Batch(Luis.LuisBatchTestClient.BatchSize);
+
+            // Evaluate each batch and add to results set
+            // We could consider running the batches in parallel
+            var batchResults = new List<ILabeledUtterance>();
+            foreach (var queryBatch in queryBatches)
+            {
+                batchResults.AddRange(await this.EvaluateAsync(queryBatch, cancellationToken).ConfigureAwait(false));
+            }
+
+            return batchResults;
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Disposes the LUIS batch test client.
+        /// </summary>
+        /// <param name="disposing">
+        /// <code>true</code> if disposing, otherwise <code>false</code>.
+        /// </param>
+        protected abstract void Dispose(bool disposing);
+
+        /// <summary>
+        /// Converts the input utterance JSON to a format suitable for LUIS batch evaluation.
+        /// </summary>
+        /// <param name="inputJson">Input utterance JSON.</param>
+        /// <returns>The modified JSON object.</returns>
+        private static JToken ToLuisBatchInput(JToken inputJson)
+        {
+            // Function to convert input entity JSON to LUIS batch format in place
+            void toLuisBatchEntity(string text, JToken json)
+            {
+                // Checks if the JSON is formatted in generic utterance format
+                if (json is JObject jsonObject && jsonObject.ContainsKey("matchText") && !jsonObject.ContainsKey("startPos"))
+                {
+                    // Extract the generic entity properties
+                    var entityType = jsonObject["entityType"];
+                    var matchText = jsonObject.Value<string>("matchText");
+                    var matchIndex = jsonObject.Value<int>("matchIndex");
+
+                    // Remove the generic entity properties
+                    jsonObject.Remove("entityType");
```

Comment:
Should we modify in place or create a new object with fields that we need to have good understanding of what is required?

Reply:
We modify in place because the batch API may support additional properties that we don't care about. For example, if the batch API started allowing users to specify a `role` property on each entity, we should pass that through. The only thing we're trying to do is remove any of the generic utterance format properties (`matchText`, `matchIndex` and `entityType`) and replace them with LUIS batch testing format. It's safe to modify in place because we do a deep clone on L129.